### PR TITLE
RD-2960

### PR DESCRIPTION
**Pull request by daniellin215**
propogate changes from java sdk (RD-2221) to js sdk (also RD-2763 from gateway proto), standingOrder and recurringTransfers.

### CODE REVIEW

File: app/src/main/RecurringTransferBuilder.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        return this;
+    }
+
+    /**
+     * Sets the amount per charge amount of the recurring payment.
+     *
+     * @param amount
+     * @return RecurringTransferTokenBuilder
+     */
+    setAmount(amount: number | string): RecurringTransferTokenBuilder {
```
Comment (zixunjian) [inline app/src/main/RecurringTransferBuilder.js]: this is taken care of by the creation function parameter right?

File: core/src/http/AuthHttpClient.js [hunk 2]
```diff
@@ -576,6 +576,43 @@ export class AuthHttpClient {
...
+    /**
+     * get information about a particular standing order
+     *
+     * @param {*} accountId
+     * @param {*} standingOrderId
+     * @param {*} keyLevel
+     * @returns
+     * @memberof AuthHttpClient
+     */
+    async getStandingOrder(accountId, standingOrderId, keyLevel) {
```
Comment (zixunjian) [inline core/src/http/AuthHttpClient.js]: move these two before `tokenOperationSignature` because that one isn't an API call.

File: core/src/http/AuthHttpClient.js [hunk 3]
```diff
@@ -576,6 +576,43 @@ export class AuthHttpClient {
         };
     }
 
+    /**
+     * get information about a particular standing order
```
Comment (zixunjian) [inline core/src/http/AuthHttpClient.js]: Comment format should be consistent (capitalization, and period)

File: core/src/http/AuthHttpClient.js [hunk 4]
```diff
@@ -576,6 +576,43 @@ export class AuthHttpClient {
...
+     *
+     * @param {*} accountId
```
Comment (zixunjian) [inline core/src/http/AuthHttpClient.js]: we should specify the type as much as we can because we don't use flow in this file

File: core/src/main/Account.js [hunk 5]
```diff
@@ -136,7 +136,7 @@ export class Account {
         return Util.callAsync(this.getTransaction, async () => {
             const res = await this.accountMember._client
                 .getTransaction(this.account.id, transactionId, keyLevel);
-            if (res.data.status !== 'SUfCCESSFUL_REQUEST')
+            if (res.data.status !== 'SUCCESSFUL_REQUEST')
```
Comment (zixunjian) [inline core/src/main/Account.js]: these should be updated like it's done in Java

File: core/src/main/Account.js [hunk 6]
```diff
@@ -166,6 +166,52 @@ export class Account {
...
+     * @return standing order record
+     */
+    getStandingOrder(
+        standingOrderId: string,
+        keyLevel: KeyLevel,
+    ): Promise<StandingOrder> {
+        return Util.callAsync(this.getStandingOrder, async () => {
+            const res = await this.accountMember._client
+                .getStandingOrder(this.account.id, standingOrderId, keyLevel);
+                if (res.data.status !== 'SUCCESSFUL_REQUEST')
```
Comment (zixunjian) [inline core/src/main/Account.js]: see above, should be `MORE_SIGNATURES_NEEDED`, like in Java

File: tpp/src/main/Member.js [hunk 7]
```diff
@@ -345,6 +346,109 @@ export default class Member extends CoreMember {
...
+     * @param currency - currency to redeem
+     * @param description - optional transfer description
+     * @param destinations - transfer destinations
+     * @param refId - ID that will be set on created Recurring Transfer
+     *                Token uses this to detect duplicates
+     *                caller might use this to recognize the transfer
+     *                if param empty, transfer will have random refId
+     * @return Recurring Transfer created as a result of this redeem call
+     */
+    redeemRecurringToken(
```
Comment (zixunjian) [inline tpp/src/main/Member.js]: in java it only takes `tokenId`, we should mimic that

**Comment by zixunjian** [comment]
We should also write a few tests for the new RecurringTransferToken functionality (and getStandingOrders) once the backend changes are in to make sure it works.

**Comment by gianluca-token** [comment]
Just a heads up, Alex had some complaints about the design recently so please hold off on merging for a bit. We're going to meet to resolve them Friday--I'll lyk the result.

File: app/src/http/AuthHttpClient.js [hunk 8]
```diff
@@ -353,6 +353,25 @@ class AuthHttpClient extends CoreAuthHttpClient {
...
+     * Creates a recurring transfer token.
+     *
+     * @param {Object} payload - payload of the token
+     * @param {string} tokenRequestId - token request ID
+     * @return {Object} response to the API call
+     */
+    async createRecurringTransferToken(payload, tokenRequestId) {
+        const request = {
+            method: 'post',
+            url: '/tokens?type=recurringTransfer',
```
[the remark this answers is not in the document]
Comment (gianluca-token) [reply app/src/http/AuthHttpClient.js]: This function should be removed anyway. You create the token by calling `createToken`.

[the remark this answers is not in the document]
Comment (gianluca-token) [reply app/src/http/AuthHttpClient.js]: `createTransferToken` and `createAccessToken` should be marked as deprecated as well.

File: tpp/src/main/Member.js (hunk 7 above)
Comment (daniellin215) [reply tpp/src/main/Member.js]: https://github.com/tokenio/sdk-java/blob/b90a866133f094c37597189dbc8b287872978333/tpp/src/main/java/io/token/tpp/Member.java#L240

The Java version has constructors that seems to account for all scenarios (number of params included in addition to tokenId). Can I make it so that they are optional, as denoted with the flow syntax with (?) ?

File: app/package.json [hunk 9]
```diff
@@ -19,7 +19,7 @@
         "build:cjs": "rollup -c",
         "build:esm": "FORMAT=esm rollup -c",
         "build:iife": "FORMAT=iife rollup -c",
-        "build": "eslint src && flow check && npm run build:cjs && npm run build:esm && npm run build:iife",
```
Comment (zixunjian) [inline app/package.json]: remember to add it back

File: app/package.json [hunk 10]
```diff
@@ -68,13 +68,14 @@
...
         "@babel/runtime-corejs2": "^7.3.1",
-        "@token-io/core": "^1.0.0",
+        "@token-io/core": "^1.0.1",
         "axios": "^0.19.0",
         "base64-js": "^1.3.0",
         "bs58": "4.0.0",
         "buffer": "^5.2.1",
         "fast-json-stable-stringify": "^2.0.0",
         "fast-sha256": "^1.1.0",
+        "lodash.merge": "^4.6.2",
```
Comment (zixunjian) [inline app/package.json]: do we need this dependency? in general we should avoid adding dependencies

File: app/sample/GetStandingOrderSubmissionsSample.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+ * @param {Member} payer - member
+ * @param {String} offset - offset
+ * @param {Number} limit - limit
+ * @return {Object} standing order submission
+ */
+export default async (payer, offset, limit) => {
+    const pagedResult = await payer.getStandingOrderSubmissions(offset, limit);
+    console.log('res', pagedResult)
+    return pagedResult.submissions;
+};
```
Comment (zixunjian) [inline app/sample/GetStandingOrderSubmissionsSample.js]: new line

File: app/sample/GetStandingOrdersSample.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    const accounts = await payer.getAccounts();
+    const accountId = accounts[0].id();
+
+    const pagedResult = await payer.getStandingOrders(
+        accountId,
+        '',
+        10,
+        config.KeyLevel.STANDARD);
+    return pagedResult.standingOrders;
+};
```
Comment (zixunjian) [inline app/sample/GetStandingOrdersSample.js]: same here

File: app/src/http/AuthHttpClient.js [hunk 13]
```diff
@@ -3,6 +3,7 @@ import Util from '../Util';
 import config from '../config.json';
 import BlockingAdapter from './BlockingAdapter';
 import base64js from 'base64-js';
+import { deprecate } from 'util';
```
Comment (zixunjian) [inline app/src/http/AuthHttpClient.js]: what is this?

File: app/src/http/AuthHttpClient.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (zixunjian) [reply app/src/http/AuthHttpClient.js]: we should remove this like luke said

File: app/src/main/AccessTokenBuilder.js [hunk 14]
```diff
@@ -74,6 +74,36 @@ export default class AccessTokenBuilder extends TokenBuilder {
         return this;
     }
 
+    /**
+     * Grants access to an account's standing orders.
+     *
+     * @param accountId
+     * @return AccessTokenBuilder
+     */
+    forAccountStandingOrder(accountId: string): AccessTokenBuilder {
```
Comment (zixunjian) [inline app/src/main/AccessTokenBuilder.js]: `forAccountStandingOrders`

File: app/src/main/AccessTokenBuilder.js [hunk 15]
```diff
@@ -74,6 +74,36 @@ export default class AccessTokenBuilder extends TokenBuilder {
...
+        return this;
+    }
+
+    /**
+     * Grants access to multiple account's standing orders.
+     *
+     * @param accountIds
+     * @returns AccessTokenBuilder
+     */
+    forAccountStandingOrders(accountIds: Array<string>): AccessTokenBuilder {
```
Comment (zixunjian) [inline app/src/main/AccessTokenBuilder.js]: `forAccountsStandingOrders`, the plural should be on accounts

File: app/src/main/Member.js [hunk 16]
```diff
@@ -361,6 +363,49 @@ export default class Member extends CoreMember {
         });
     }
 
+    /**
+     * Creates a new standing order token builder. Defines a standing order
+     * for a fixed time span.
+     *
+     * @param {(number | string)} amount individual transfer amount
```
Comment (zixunjian) [inline app/src/main/Member.js]: we dont need types here since we have flow in this file.

File: tpp/package.json [hunk 17]
```diff
@@ -15,7 +15,7 @@
     "scripts": {
         "lint": "eslint .",
         "test": "npm run build && npm run testOnly",
-        "testOnly": "eslint test && mocha-webpack --webpack-config test/webpack.config.js --full-trace --check-leaks -t 60000 -c \"test/**/*.spec.js\" && rimraf keys",
+        "testOnly": "mocha-webpack --webpack-config test/webpack.config.js --full-trace --check-leaks -t 60000 -c \"test/sample/RedeemStandingOrderTokenSample.spec.js\" && rimraf keys",
```
Comment (zixunjian) [inline tpp/package.json]: revert this

File: tpp/src/main/Member.js [hunk 18]
```diff
@@ -265,30 +266,30 @@ export default class Member extends CoreMember {
...
+     *                If param empty, transfer will have random refId.
      * @return Transfer created as a result of this redeem call
      */
     redeemToken(
         token: Token | string,
-        amount?: number,
-        currency?: string,
-        description?: string,
-        destinations?: Array<TransferDestination> = [],
+        amount: ?number,
```
Comment (zixunjian) [inline tpp/src/main/Member.js]: why this change?

File: tpp/src/main/Member.js [hunk 19]
```diff
@@ -265,30 +266,30 @@ export default class Member extends CoreMember {
...
+        currency: ?string,
+        description: ?string,
+        destinations: Array<TransferDestination> = [],
         refId?: string
     ): Promise<Transfer> {
         return Util.callAsync(this.redeemToken, async () => {
             const finalToken = await this._resolveToken(token);
             if (!amount) {
-                amount = finalToken.payload.transfer.lifetimeAmount;
+                amount = finalToken && finalToken.payload.transfer.lifetimeAmount;
```
Comment (zixunjian) [inline tpp/src/main/Member.js]: do we need this check here?

File: tpp/test/TestUtil.js [hunk 20]
```diff
@@ -55,7 +55,31 @@ export default class TestUtil {
             .setToAlias(payeeAlias)
             .setDescription('Book Purchase')
             .execute();
+        console.log(transferToken);
```
Comment (zixunjian) [inline tpp/test/TestUtil.js]: rm

File: tpp/test/TestUtil.js [hunk 21]
```diff
@@ -55,7 +55,31 @@ export default class TestUtil {
...
     }
+
+    static async createStandingOrderToken(payer, payeeAlias) {
+        const accountId = (await payer.getAccounts())[0].id();
+        const destination = {
+                sepa: {
+                    iban: '123',
+                },
+        };
+        console.log('here', accountId)
```
Comment (zixunjian) [inline tpp/test/TestUtil.js]: this function has lots of log statements and should be cleaned up

File: tpp/test/webpack.config.js [hunk 22]
```diff
@@ -1,6 +1,7 @@
 const webpack = require('webpack');
 const nodeExternals = require('webpack-node-externals');
 const {resolve, dirname} = require('path');
+const axios = require('axios');
```
Comment (zixunjian) [inline tpp/test/webpack.config.js]: is this used?

File: app/src/http/AuthHttpClient.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (daniellin215) [reply app/src/http/AuthHttpClient.js]: should I also remove createTransferToken and createAccessToken?

File: tpp/src/main/Member.js (hunk 19 above)
Comment (daniellin215) [reply tpp/src/main/Member.js]: i reverted, don't remember making these changes

File: tpp/test/webpack.config.js (hunk 22 above)
Comment (daniellin215) [reply tpp/test/webpack.config.js]: Not sure, don't remember adding this

File: app/package.json (hunk 10 above)
Comment (daniellin215) [reply app/package.json]: Had to add this to resolve npm error, security vulnerability before

File: app/src/http/AuthHttpClient.js (hunk 13 above)
Comment (daniellin215) [reply app/src/http/AuthHttpClient.js]: rm, auto import when I used deprecate in comments
